### PR TITLE
Add preview header for newly accessible endpoints

### LINKED ISSUE #722: Preview of additional endpoints for GitHub Apps.
GitHub Developer Announcement:
https://developer.github.com/changes/2017-09-21-additional-endpoints-for-github-apps/

Woohoo!  Lots of new endpoints to support.

I'm thinking that this is way too big for a single PR, and yet it would probably be silly to open up a separate issue for each endpoint... so how about if I add a checkbox task list to the endpoints in this issue and we can check them off as PRs address them? Obviously, the first PR will also need to add the new custom media type `application/vnd.github.machine-man-preview+json` in the `Accept` header and following PRs can just use it.

Here are the recently-added endpoints:
- [ ] [GET /](https://developer.github.com/v3/#root-endpoint)
- [ ] [GET /emojis](https://developer.github.com/v3/emojis/)
- [ ] [GET /events](https://developer.github.com/v3/activity/events/#list-public-events)
- [ ] [GET /networks/:owner/:repo/events](https://developer.github.com/v3/activity/events/#list-public-events-for-a-network-of-repositories)
- [ ] [GET /orgs/:org/events](https://developer.github.com/v3/activity/events/#list-public-events-for-an-organization)
- [ ] [GET /users/:username/events](https://developer.github.com/v3/activity/events/#list-events-performed-by-a-user)
- [ ] [GET /users/:username/events/public](https://developer.github.com/v3/activity/events/#list-public-events-performed-by-a-user)
- [ ] [GET /users/:username/received_events](https://developer.github.com/v3/activity/events/#list-events-that-a-user-has-received)
- [ ] [GET /users/:username/received_events/public](https://developer.github.com/v3/activity/events/#list-public-events-that-a-user-has-received)
- [ ] [GET /repos/:owner/:name/community/profile](https://developer.github.com/v3/repos/community/#retrieve-community-profile-metrics)
- [ ] [GET /repos/:owner/:repo/community/code_of_conduct](https://developer.github.com/v3/codes_of_conduct/#get-the-contents-of-a-repositorys-code-of-conduct)
- [ ] [GET /repos/:owner/:repo/branches/:branch/protection](https://developer.github.com/v3/repos/branches/#get-branch-protection)
- [ ] [GET /repos/:owner/:repo/branches/:branch/protection/restrictions](https://developer.github.com/v3/repos/branches/#get-restrictions-of-protected-branch)
- [ ] [GET /repos/:owner/:repo/branches/:branch/protection/restrictions/users](https://developer.github.com/v3/repos/branches/#list-user-restrictions-of-protected-branch)
- [ ] [GET /repos/:owner/:repo/branches/:branch/protection/restrictions/teams](https://developer.github.com/v3/repos/branches/#list-team-restrictions-of-protected-branch)
- [ ] [GET /repos/:owner/:repo/branches/:branch/protection/required_pull_request_reviews](https://developer.github.com/v3/repos/branches/#get-pull-request-review-enforcement-of-protected-branch)
- [ ] [GET /repos/:owner/:repo/branches/:branch/protection/enforce_admins](https://developer.github.com/v3/repos/branches/#get-admin-enforcement-of-protected-branch)
- [ ] [PUT /repos/:owner/:repo/branches/:branch/protection](https://developer.github.com/v3/repos/branches/#update-branch-protection)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection](https://developer.github.com/v3/repos/branches/#remove-branch-protection)
- [ ] [PATCH /repos/:owner/:repo/branches/:branch/protection/required_status_checks](https://developer.github.com/v3/repos/branches/#update-required-status-checks-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/required_status_checks](https://developer.github.com/v3/repos/branches/#remove-required-status-checks-of-protected-branch)
- [ ] [POST /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts](https://developer.github.com/v3/repos/branches/#add-required-status-checks-contexts-of-protected-branch)
- [ ] [PUT /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts](https://developer.github.com/v3/repos/branches/#replace-required-status-checks-contexts-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts](https://developer.github.com/v3/repos/branches/#remove-required-status-checks-contexts-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/restrictions](https://developer.github.com/v3/repos/branches/#remove-restrictions-of-protected-branch)
- [ ] [PUT /repos/:owner/:repo/branches/:branch/protection/restrictions/users](https://developer.github.com/v3/repos/branches/#replace-user-restrictions-of-protected-branch)
- [ ] [POST /repos/:owner/:repo/branches/:branch/protection/restrictions/users](https://developer.github.com/v3/repos/branches/#add-user-restrictions-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/restrictions/users](https://developer.github.com/v3/repos/branches/#remove-user-restrictions-of-protected-branch)
- [ ] [PUT /repos/:owner/:repo/branches/:branch/protection/restrictions/teams](https://developer.github.com/v3/repos/branches/#replace-team-restrictions-of-protected-branch)
- [ ] [POST /repos/:owner/:repo/branches/:branch/protection/restrictions/teams](https://developer.github.com/v3/repos/branches/#add-team-restrictions-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/restrictions/teams](https://developer.github.com/v3/repos/branches/#remove-team-restrictions-of-protected-branch)
- [ ] [PATCH /repos/:owner/:repo/branches/:branch/protection/required_pull_request_reviews](https://developer.github.com/v3/repos/branches/#update-pull-request-review-enforcement-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/required_pull_request_reviews](https://developer.github.com/v3/repos/branches/#remove-pull-request-review-enforcement-of-protected-branch)
- [ ] [POST /repos/:owner/:repo/branches/:branch/protection/enforce_admins](https://developer.github.com/v3/repos/branches/#add-admin-enforcement-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/enforce_admins](https://developer.github.com/v3/repos/branches/#remove-admin-enforcement-of-protected-branch)

Here are the endpoints for [user-to-server requests](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/identifying-users-for-github-apps/#user-to-server-requests):

- [ ] [POST /repos/:owner/:repo/git/trees](https://developer.github.com/v3/git/trees/#create-a-tree)
- [ ] [POST /repos/:owner/:repo/git/refs](https://developer.github.com/v3/git/refs/#create-a-reference)
- [ ] [POST /repos/:owner/:repo/git/blobs](https://developer.github.com/v3/git/blobs/#create-a-blob)
- [ ] [POST /repos/:owner/:repo/git/commits](https://developer.github.com/v3/git/commits/#create-a-commit)
- [ ] [POST /repos/:owner/:repo/git/tags](https://developer.github.com/v3/git/tags/#create-a-tag-object)
- [ ] [PATCH /repos/:owner/:repo/git/refs/:ref](https://developer.github.com/v3/git/refs/#update-a-reference)
- [ ] [DELETE /repos/:owner/:repo/git/refs/:ref](https://developer.github.com/v3/git/refs/#delete-a-reference)
- [ ] [PUT /repos/:owner/:repo/topics](https://developer.github.com/v3/repos/#replace-all-topics-for-a-repository)
- [ ] [PUT /repos/:owner/:repo/contents/:path](https://developer.github.com/v3/repos/contents/#create-a-file) (create)
- [ ] [PUT /repos/:owner/:repo/contents/:path](https://developer.github.com/v3/repos/contents/#update-a-file) (update)
- [ ] [DELETE /repos/:owner/:repo/contents/:path](https://developer.github.com/v3/repos/contents/#delete-a-file)
- [ ] [POST /repos/:owner/:repo/statuses/:sha](https://developer.github.com/v3/repos/statuses/#create-a-status)
- [ ] [PUT /repos/:owner/:repo/branches/:branch/protection](https://developer.github.com/v3/repos/branches/#update-branch-protection)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection](https://developer.github.com/v3/repos/branches/#remove-branch-protection)
- [ ] [PATCH /repos/:owner/:repo/branches/:branch/protection/required_status_checks](https://developer.github.com/v3/repos/branches/#update-required-status-checks-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/required_status_checks](https://developer.github.com/v3/repos/branches/#remove-required-status-checks-of-protected-branch)
- [ ] [POST /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts](https://developer.github.com/v3/repos/branches/#add-required-status-checks-contexts-of-protected-branch)
- [ ] [PUT /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts](https://developer.github.com/v3/repos/branches/#replace-required-status-checks-contexts-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts](https://developer.github.com/v3/repos/branches/#remove-required-status-checks-contexts-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/restrictions](https://developer.github.com/v3/repos/branches/#remove-restrictions-of-protected-branch)
- [ ] [PUT /repos/:owner/:repo/branches/:branch/protection/restrictions/users](https://developer.github.com/v3/repos/branches/#replace-user-restrictions-of-protected-branch)
- [ ] [POST /repos/:owner/:repo/branches/:branch/protection/restrictions/users](https://developer.github.com/v3/repos/branches/#add-user-restrictions-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/restrictions/users](https://developer.github.com/v3/repos/branches/#remove-user-restrictions-of-protected-branch)
- [ ] [PUT /repos/:owner/:repo/branches/:branch/protection/restrictions/teams](https://developer.github.com/v3/repos/branches/#replace-team-restrictions-of-protected-branch)
- [ ] [POST /repos/:owner/:repo/branches/:branch/protection/restrictions/teams](https://developer.github.com/v3/repos/branches/#add-team-restrictions-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/restrictions/teams](https://developer.github.com/v3/repos/branches/#remove-team-restrictions-of-protected-branch)
- [ ] [PATCH /repos/:owner/:repo/branches/:branch/protection/required_pull_request_reviews](https://developer.github.com/v3/repos/branches/#update-pull-request-review-enforcement-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/required_pull_request_reviews](https://developer.github.com/v3/repos/branches/#remove-pull-request-review-enforcement-of-protected-branch)
- [ ] [POST /repos/:owner/:repo/branches/:branch/protection/enforce_admins](https://developer.github.com/v3/repos/branches/#add-admin-enforcement-of-protected-branch)
- [ ] [DELETE /repos/:owner/:repo/branches/:branch/protection/enforce_admins](https://developer.github.com/v3/repos/branches/#remove-admin-enforcement-of-protected-branch)
- [ ] [PATCH /repos/:owner/:repo/issues/:number](https://developer.github.com/v3/issues/#edit-an-issue)

It seems reasonable to assign multiple owners to this issue, and here in the comments we will attempt to clearly document who is working on which endpoints so that we don't waste anyone's time or effort.

Thank you in advance for contributing to this open source project!
Your assistance is greatly appreciated.

### PR DESCRIPTION
**Pull request by elliott-beach**
Fixes #722. This will add the preview header for the endpoints newly accessible to GitHub apps.

### CODE REVIEW

**Comment by e-beach** [comment]
Skipping `/repos/:owner/:repo/branches/:branch/protection/restrictions` because the endpoint is not currently supported by us.

**Comment by e-beach** [comment]
@shurcooL Would you mind reviewing the PR at this point? Adding the headers to more endpoints is essentially mechanical from here, so I'd rather be doing things exactly right before I add 20 more endpoints. 

Some points I'm curious on:

- I added a `testHeaders` method to avoid some annoying code repetition, does this make sense?
- The style of the `TODO` comments. I could note, especially when there are multiple types in the Accept header, that `mediaTypeGithubAppsPreview` does not mean an endpoint is in preview. This might ease maintenance down the road by clarifying when to remove an accept type.

File: github/repos.go [hunk 1]
```diff
@@ -320,8 +320,9 @@ func (s *RepositoriesService) GetCodeOfConduct(ctx context.Context, owner, repo
 		return nil, nil, err
 	}
 
-	// TODO: remove custom Accept header when this API fully launches.
-	req.Header.Set("Accept", mediaTypeCodesOfConductPreview)
+	// TODO: remove custom Accept header when this API fully launches
+	acceptHeaders := []string{mediaTypeCodesOfConductPreview, mediaTypeGitHubAppsPreview}
+	req.Header.Set("Accept", strings.Join(acceptHeaders, ","))
```
Comment (shurcooL) [inline github/repos.go]: Up until this PR, we've been using `", "` (comma, followed by space) rather than `","` (just comma) as the header value separator... I think we should be consistent.

Did you want to drop the space on purpose? If it's valid (most likely, but should be tested if we don't want to take any chances), I suppose it's a little more space efficient, at the cost of being slightly less human readable.

In any case, given how big and hard to deal with this PR will be already, I'd propose you factor out this change into a separate PR, and keep this one as minimal as possible. So, use comma-space as before for consistency.

**Comment by shurcooL** [comment]
See inline comment about the accept header.

Aside from that, I think this looks good.

However, I will say that I will be quite afraid of merging a PR that will touch accept headers of so many endpoints at once. I think it would be great if we actually tested out the final PR before merging it in any applications we have access to, just as as a sanity check.

> I added a `testHeaders` method to avoid some annoying code repetition, does this make sense?

I checked to see if it was really worth including (i.e., I compared to what the code would be without it), by trying to rewrite the code without it:

```diff
-testHeaders(t, r, "Accept", []string{mediaTypeCodesOfConductPreview, mediaTypeGitHubAppsPreview})
+testHeader(t, r, "Accept", strings.Join([]string{mediaTypeCodesOfConductPreview, mediaTypeGitHubAppsPreview}, ", "))
```

In the end, I think it's fair to add it.

> The style of the `TODO` comments. I could note, especially when there are multiple types in the Accept header, that `mediaTypeGithubAppsPreview` does not mean an endpoint is in preview. This might ease maintenance down the road by clarifying when to remove an accept type.

Honestly, given the pattern I'm seeing the last few months of GitHub seemingly adding more and more things to preview headers, quite a few overlapping, and this issue included, I'm starting to think that comment has outlived its purpose and we might as well stop caring about including it for consistency.

In other words, at this point, I think we recognize the pattern of setting custom preview (or otherwise) accept headers even without the TODO comment, so I don't have any strong feelings about what to do with it. If you do or don't include it, that'd be okay with me. So go with whatever you think makes most sense, or how @gmlewis feels about it.

File: github/repos.go (hunk 1 above)
Comment (e-beach) [reply github/repos.go]: Thanks! I'll just use `", "` for consistency.

**Comment by e-beach** [comment]
OK, I agree that the TODO comment is boilerplate and am going to drop it.

**Comment by e-beach** [comment]
I noticed some endpoints on GiHub's list are not supported by our API.

```
PATCH /repos/:owner/:repo/branches/:branch/protection/required_status_checks
DELETE /repos/:owner/:repo/branches/:branch/protection/required_status_checks
POST /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts
PUT /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts
DELETE /repos/:owner/:repo/branches/:branch/protection/required_status_checks/contexts
GET /repos/:owner/:repo/branches/:branch/protection/restrictions
DELETE /repos/:owner/:repo/branches/:branch/protection/restrictions
PUT /repos/:owner/:repo/branches/:branch/protection/restrictions/users
POST /repos/:owner/:repo/branches/:branch/protection/restrictions/users
DELETE /repos/:owner/:repo/branches/:branch/protection/restrictions/users
PUT /repos/:owner/:repo/branches/:branch/protection/restrictions/teams
POST /repos/:owner/:repo/branches/:branch/protection/restrictions/teamsDELETE /repos/:owner/:repo/branches/:branch/protection/restrictions/teams
```

**Comment by e-beach** [comment]
What remains are the user-to-server requests. However, my enthusiasm for adding the preview headers has waned, so I suggest we add those in a separate PR.

File: github/github.go [hunk 2]
```diff
@@ -102,6 +102,9 @@ const (
 
 	// https://developer.github.com/changes/2017-07-26-team-review-request-thor-preview/
 	mediaTypeTeamReviewPreview = "application/vnd.github.thor-preview+json"
+
+	// https://developer.github.com/changes/2017-09-21-additional-endpoints-for-github-apps/
+	mediaTypeGitHubAppsPreview = "application/vnd.github.machine-man-preview+json"
```
Comment (shurcooL) [inline github/github.go]: I just noticed we already have the same custom media type constant defined above:

```Go
// https://developer.github.com/changes/2016-09-14-Integrations-Early-Access/
mediaTypeIntegrationPreview = "application/vnd.github.machine-man-preview+json"
```

**Comment by elliott-beach** [comment]
Yes, you’re right that this is not very important. I always appreciate the detailed arguments you put forth, thanks!

I’ll close this on that basis.
